src/GovUk.Education.ExploreEducationStatistics.Content.Api/Controllers/ReleaseController.cs

```diff
@@ -35,39 +43,71 @@ public async Task<ActionResult<List<ReleaseSummaryViewModel>>> ListReleases(stri
                 .HandleFailuresOrOk();
         }
 
-        [MemoryCache(typeof(GetLatestReleaseCacheKey), durationInSeconds: 10, expiryScheduleCron: HalfHourlyExpirySchedule)]
+        [MemoryCache(typeof(GetLatestReleaseCacheKey), durationInSeconds: 10,
+            expiryScheduleCron: HalfHourlyExpirySchedule)]
         [HttpGet("publications/{publicationSlug}/releases/latest")]
         public async Task<ActionResult<ReleaseViewModel>> GetLatestRelease(string publicationSlug)
         {
-            return await _releaseCacheService.GetReleaseAndPublication(publicationSlug)
+            return await GetReleaseViewModel(publicationSlug)
                 .HandleFailuresOrOk();
         }
 
         [HttpGet("publications/{publicationSlug}/releases/latest/summary")]
         public async Task<ActionResult<ReleaseSummaryViewModel>> GetLatestReleaseSummary(string publicationSlug)
         {
-            return await _releaseCacheService.GetReleaseSummary(publicationSlug)
+            return await GetReleaseSummaryViewModel(publicationSlug)
                 .HandleFailuresOrOk();
         }
 
         [MemoryCache(typeof(GetReleaseCacheKey), durationInSeconds: 15, expiryScheduleCron: HalfHourlyExpirySchedule)]
         [HttpGet("publications/{publicationSlug}/releases/{releaseSlug}")]
         public async Task<ActionResult<ReleaseViewModel>> GetRelease(string publicationSlug, string releaseSlug)
         {
-            return await _releaseCacheService.GetReleaseAndPublication(
-                    publicationSlug,
-                    releaseSlug)
+            return await GetReleaseViewModel(publicationSlug, releaseSlug)
                 .HandleFailuresOrOk();
         }
 
         [HttpGet("publications/{publicationSlug}/releases/{releaseSlug}/summary")]
         public async Task<ActionResult<ReleaseSummaryViewModel>> GetReleaseSummary(string publicationSlug,
             string releaseSlug)
         {
-            return await _releaseCacheService.GetReleaseSummary(
+            return await GetReleaseSummaryViewModel(
                     publicationSlug,
                     releaseSlug)
                 .HandleFailuresOrOk();
         }
+
+        private Task<Either<ActionResult, ReleaseViewModel>> GetReleaseViewModel(
+            string publicationSlug,
+            string? releaseSlug = null)
+        {
+            return _publicationCacheService.GetPublication(publicationSlug)
+                .OnSuccessCombineWith(publication => _methodologyCacheService.GetSummariesByPublication(publication.Id))
+                .OnSuccessCombineWith(_ => _releaseCacheService.GetRelease(publicationSlug, releaseSlug))
```

Comment:
Not saying that we need to do this, but did we consider optimising this to make parallel requests rather than performing these sequentially?

Reply:
I don't think we can easily do this for two reasons.

1. [EES-3516](https://dfedigital.atlassian.net/browse/EES-3516), mentions that `CacheAspect` is forcing all `CacheAttribute` annotated methods to run synchronously.
2. EF Core doesn't support multiple parallel operations being run on the same context instance. This would be a problem if two or more methods got cache-misses and run the underlying methods which do database work. We need to await each one before starting the next.

I commented out the `MemoryCache` annotation on the method calling this one to make sure this block always runs, and then did some investigations.

I also ignores the third task getting methodologies for simplicity and just work with two tasks.

### Notes on (1):

A task getting a result from a cache is run  synchronously by
https://github.com/dfe-analytical-services/explore-education-statistics/blob/dev/src/GovUk.Education.ExploreEducationStatistics.Common/Cache/CacheAspect.cs#L81 :

`var triggerResult = cacheTrigger.Get(cacheKey, cachedReturnType).Result;`

If there's a cache-miss, the task getting the underlying result is run synchronously by https://github.com/dfe-analytical-services/explore-education-statistics/blob/dev/src/GovUk.Education.ExploreEducationStatistics.Common/Extensions/ReflectionExtensions.cs#L120 :

`task.Wait()`

So if I start two tasks as follows without awaiting them, we will see their status is always already complete and we get the log message "Tasks already ran to completion":

```
var publicationTask = _publicationCacheService.GetPublication(publicationSlug);
var releaseTask = _releaseCacheService.GetRelease(publicationSlug, releaseSlug);

if (publicationTask.Status == TaskStatus.RanToCompletion
    && releaseTask.Status == TaskStatus.RanToCompletion)
{
    _logger.LogInformation("Tasks already ran to completion");
}
```

We see the same result whether there's a cache-miss or not. The cache-misses are reproduced by clearing the cache before running. 

### Notes on (2):

We can force parallel running by commenting out the `BlobCache` annotation on `GetPublication` and `GetRelease` in the cache services.

```
    //[BlobCache(typeof(PublicationCacheKey), ServiceName = "public")]
    public Task<Either<ActionResult, PublicationCacheViewModel>> GetPublication(string publicationSlug)

    //[BlobCache(typeof(ReleaseCacheKey), ServiceName = "public")]
    public Task<Either<ActionResult, ReleaseCacheViewModel>> GetRelease(string publicationSlug, string? releaseSlug = null)
```

We then run into the database context blocking issue.

Here's the code running them in parallel:

```
var publicationTask = _publicationCacheService.GetPublication(publicationSlug);
var releaseTask = _releaseCacheService.GetRelease(publicationSlug, releaseSlug);

if (publicationTask.Status == TaskStatus.RanToCompletion
    && releaseTask.Status == TaskStatus.RanToCompletion)
{
    _logger.LogInformation("Tasks already ran to completion");
    return new ReleaseViewModel(
        releaseTask.Result.Right,
        new PublicationViewModel(publicationTask.Result.Right,
            new List<MethodologyVersionSummaryViewModel>())
    );
}

// Tasks are running in parallel
_logger.LogInformation("Running async");

var publication = await publicationTask;
var release = await releaseTask;

return new ReleaseViewModel(
    release.Right,
    new PublicationViewModel(publication.Right, new List<MethodologyVersionSummaryViewModel>())
);
```

and the error

```
System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
```

Reply:
Ah bad times. Surprising behaviour that the AspectInjector wrapped tasks seem to run to completion immediately!

Oh well, happy to leave things as-is!